include/upipe/ulog.h

```diff
@@ -78,13 +82,33 @@ struct ulog {
  */
 static inline void ulog_init(struct ulog *ulog,
                              enum uprobe_log_level level,
-                             const char *msg)
+                             const char *format)
 {
     ulog->level = level;
-    ulog->msg = msg;
+    ulog->format = format;
     ulist_init(&ulog->prefixes);
 }
 
+UBASE_PRAGMA_CLANG(diagnostic push)
+UBASE_PRAGMA_CLANG(diagnostic ignored "-Wformat-nonliteral")
+
+static inline int ulog_msg_print(struct ulog *ulog, char *buf, size_t size)
+{
+    va_list args;
+
+    va_copy(args, ulog->args);
+    int ret = vsnprintf(buf, size, ulog->format, args);
+    va_end(args);
+    return ret;
+}
+
+UBASE_PRAGMA_CLANG(diagnostic pop)
```

Comment:
It would be nice to have a comment explaining why this is needed

Reply:
Done.